### PR TITLE
Add emoji lookup

### LINKED ISSUE #376: Emoji Lookup
Include emoji lookup. Similar to slack emoji use. Helps using emojis without leaving the keyboard. 

Example - ":smile:" or ":bear:"

### PR DESCRIPTION
**Pull request by FredericoJunqueira**
Added an emoji lookup that operates based on string codes such as ":smile", ":small_plan...". 

Fixes #376

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for working on this :)

**Comment by sindresorhus** [comment]
Don't do unrelated changes and run `npx xo --fix` (Ensure you have the latest Node.js version installed) locally and fix any remaining style issues.

File: browser.js [hunk 1]
```diff
@@ -183,6 +184,30 @@ ipc.on('toggle-dark-mode', () => {
...
+function setEmojiSearch(){
+	var textInput = document.querySelector('._5rpu');
+	emoji.load();
+
+	if(!textInput){
+		return;
+	}
+	
+	textInput.onkeyup = function(){
+		delay(function(){
```
Comment (sindresorhus) [inline browser.js]: Would be better to just use `lodash.debounce` here.

**Comment by FredericoJunqueira** [comment]
Ok! Later today I will replace the delay function with your suggestion and
also update node and fix all the styling errors!

Best regards,
Frederico Junqueira

**Comment by sindresorhus** [comment]
You still have lint errors: https://travis-ci.org/sindresorhus/caprine/builds/396646536?utm_source=github_status&utm_medium=notification

**Comment by sindresorhus** [comment]
The dropdown appears out of view when you have a wide window: 
screen shot 2018-06-27 at 13 05 51 It should also be absolute positioned so it doesn't move around any other elements. It should just float above the text.

This is how it looks in Slack:

screen shot 2018-06-27 at 13 07 35

The first result should be highlighted by default, and if I press enter, it should be inserted, but the message should not yet be sent.

**Comment by sindresorhus** [comment]
When I remove my text, the emoji lookup doesn't disappear. It should also disappear when I press `Esc`.

File: browser.js [hunk 2]
```diff
@@ -183,6 +185,20 @@ ipc.on('toggle-dark-mode', () => {
...
+	emoji.load();
+
+	if (!textInput) {
+		return;
+	}
+
+	textInput.addEventListener('keyup', debounce(() => {
+		const lastword = textInput.textContent.split(' ').pop();
+		emoji.parse(lastword, textInput);
+	}, 800, false));
```
Comment (sindresorhus) [inline browser.js]: 800 is too long. I would go for 400.

File: emoji.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
+// Emoji datasource
+const emoji = require('emojilib');
+// Regex expression chosen to identify the string as emoji label
+const regexExpression = /(:[\w\-+]+)/g;
+// Emoji array
```
Comment (sindresorhus) [inline emoji.js]: Use an actual `Map` here.

File: emoji.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	});
+}
+
+// Function to construct the emoji dropdown resultant from the search
+function buildDropdown(array, currentText) {
+	const list = document.createElement('ul');
+	list.className = 'textcomplete-dropdown';
+	list.id = 'emoji-options';
+
+	for (let i = 0; i < array.length; i++) {
```
Comment (sindresorhus) [inline emoji.js]: Use a `for-of` loop

File: emoji.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+		const li = document.createElement('li');
+		li.innerHTML = `<li class="textcomplete-item">${opt[0]} <span class="emoji">${opt[1].char}</span></li>`;
+
+		li.addEventListener('click', function () {
+			const text = document.querySelectorAll('[data-text="true"]')[0];
+			const emojiChar = this.getElementsByClassName('emoji')[0].innerHTML;
+			text.innerHTML = text.innerHTML.replace(currentText, emojiChar);
+
+			document.getElementById('emoji-options').remove();
```
Comment (sindresorhus) [inline emoji.js]: You already have reference to `list`. You can just use that here instead.

File: emoji.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+	for (let i = 0; i < array.length; i++) {
+		const opt = array[i];
+
+		const li = document.createElement('li');
+		li.innerHTML = `<li class="textcomplete-item">${opt[0]} <span class="emoji">${opt[1].char}</span></li>`;
+
+		li.addEventListener('click', function () {
+			const text = document.querySelectorAll('[data-text="true"]')[0];
+			const emojiChar = this.getElementsByClassName('emoji')[0].innerHTML;
```
Comment (sindresorhus) [inline emoji.js]: Use `querySelector`

File: package.json [hunk 7]
```diff
@@ -29,7 +29,9 @@
 		"electron-store": "^1.1.0",
 		"electron-updater": "^2.21.8",
 		"element-ready": "^2.2.0",
-		"facebook-locales": "^1.0.464"
+		"emojilib": "^2.2.12",
+    "facebook-locales": "^1.0.464",
+    "lodash.debounce": "^4.0.8"
```
Comment (sindresorhus) [inline package.json]: Wrong indentation

File: browser.css [hunk 8]
```diff
@@ -330,3 +330,18 @@ a,
...
+	list-style: none;
+  }
+  
+  .textcomplete-item {
+	padding: 10px;
+  }
+  
+  .textcomplete-item:not(:last-child) {
+	border-bottom: 1px solid #ccc
+  }
```
Comment (sindresorhus) [inline browser.css]: Some indentation weirdness here.

File: browser.js [hunk 9]
```diff
@@ -183,6 +185,20 @@ ipc.on('toggle-dark-mode', () => {
 	setDarkMode();
 });
 
+function setEmojiSearch() {
```
Comment (sindresorhus) [inline browser.js]: I think this logic should be in `emoji.js` too. All you need in this file is `initEmoji(textInput);`

**Comment by FredericoJunqueira** [comment]
Hi Sindre,

I'm currently working on the changes you've proposed. Most of them are
done, but I'm still struggling on placing the emoji selector above the chat
window (on top of the text editor). I'm also working on the actions (enter,
esc, move). Do you suggest using the regex pattern inside a map object?

Best regards,
Frederico Junqueira

File: emoji.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+let emojis = null;
+
+// Function to load all the emojis from the data source
+function load() {
+	if (emojis) {
+		return emojis;
+	}
+
+	const result = emoji.lib;
+	emojis = Object.keys(result).map(key => {
```
Comment (sindresorhus) [inline emoji.js]: Use `Object.entries()` here instead of `Object.keys()`

File: emoji.js [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+	}, 400, false));
+
+	document.addEventListener('keyup', e => {
+		const picker = document.querySelector('.emoji-menu');
+		if (picker !== undefined) {
+			e.stopImmediatePropagation();
+			e.stopPropagation();
+			e.preventDefault();
+
+			switch (e.keyCode) {
```
Comment (sindresorhus) [inline emoji.js]: Use `e.key` instead

See: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

File: emoji.js [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+// Function to check if search string is part of an emoji label
+function exist(emojiId) {
+	const emojiMap = emojis;
+
+	if (emojiMap === null) {
+		return false;
+	}
+
+	return emojiMap.some(k => {
+		return k[0].indexOf(emojiId) !== -1;
```
Comment (sindresorhus) [inline emoji.js]: Use `Array#includes`

**Comment by sindresorhus** [comment]
> but I'm still struggling on placing the emoji selector above the chat
window (on top of the text editor)

https://github.com/component/textarea-caret-position

> Do you suggest using the regex pattern inside a map object?


Not sure I understand what you're asking.

**Comment by FredericoJunqueira** [comment]
I'm still working on cancelling the message submit, when pressing enter with the emoji picker open.

**Comment by dangkhoa141** [comment]
Would love to see this PR get merged soon. Great job @FredericoJunqueira

**Comment by Ginkooo** [comment]
@FredericoJunqueira I'm to lazy to send likes by clicking with mouse, please finish that PR, people need you!

**Comment by sindresorhus** [comment]
Closing for lack of activity.

**Pull request closed without merging**